Allow favoriting instance groups

aoldershaw opened the pull request:
<!--
Hi there! Thanks for submitting a pull request to Concourse!

The title of your pull request will be used to generate the release notes.
Please provide a brief sentence that describes the PR, using the [imperative
mood]. Please refrain from adding prefixes like 'feature:', and don't include a
period at the end.

Examples: "Add feature to doohickey", "Fix panic during spline reticulation"

We will edit the title if needed so don't worry about getting it perfect!

To help us review your PR, please fill in the following information.
-->

[imperative mood]: https://chris.beams.io/posts/git-commit/#imperative

## What does this PR accomplish?
<!--
Choose all that apply.
Also, mention the linked issue here.
This will magically close the issue once the PR is merged.
-->
Bug Fix | **Feature** | Documentation

<img width="1000" alt="Screen Shot 2021-01-23 at 9 41 49 PM" src="https://user-images.githubusercontent.com/26583442/105619802-114a7e80-5dc4-11eb-8f66-614d696fa9d0.png">

Previously, only individual instances could be favourited. When instances were favourited, the instance group card would be displayed in the favourites section, but with the set of instances filtered down to only those that were favourited.

I figured you'd typically only care about a few instances, and the old approach made it easy to find those instances at a glance - but there are definitely valid cases for wanting to see all instances at once, and this PR makes that possible (without needing to favourite each individual instance).

## Changes proposed by this PR:
<!--
Tell the reviewer What changed, Why, and How were you able to accomplish that?
-->

* Store favourited instance groups in `localStorage`
* Store sidebar state in `localStorage` (rather than `sessionStorage`)
  * Wasn't strictly necessary, but allowed simplifying the code, and if you use the sidebar, you'll probably want it to stay open when you open a new tab
* Display `PipelineCard` for each favourited instance, even when not viewing instance groups via the `group:...` filter
  * The `PipelineCards` contain both the group name + instance vars
  * In contrast, when viewing instance groups, the `PipelineCards` only contain the instance vars, because the headers outside of the cards contain the instance group name
* Display pipeline list items in the sidebar for favourited instances using the same `fly` notation
  * i.e. `group-name/var1:val1,var2:val2`

## Notes to reviewer:
<!--
Leave a message to whoever is going to review this PR.
Mainly, pointers to review the PR, and how they can test it.
-->

## Contributor Checklist
<!--
Most of the PRs should have the following added to them,
this doesn't apply to all PRs, so it is helpful to tell us what you did.
-->
- [ ] Followed [Code of conduct], [Contributing Guide] & avoided [Anti-patterns]
- [ ] [Signed] all commits
- [ ] Added tests (Unit and/or Integration)
- [ ] Updated [Documentation]
- [ ] Added release note (Optional)

[Code of Conduct]: https://github.com/concourse/concourse/blob/master/CODE_OF_CONDUCT.md
[Contributing Guide]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md
[Anti-patterns]: https://github.com/concourse/concourse/wiki/Anti-Patterns
[Signed]: https://help.github.com/en/github/authenticating-to-github/signing-commits
[Documentation]: https://github.com/concourse/docs

## Reviewer Checklist
<!--
This section is intended for the reviewers only, to track review
progress.
-->
- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
- [ ] New config flags added? Ensure that they are added to the
  [BOSH](https://github.com/concourse/concourse-bosh-release) and
  [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for
  the [integration
  tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env)
  (for example, if they are Garden configs that are not displayed in the
  `--help` text).

matthewpereira commented:
One idea to think about and get your opinion on @aoldershaw:

In the design studio for this feature, folks seemed to like representing the 'tree structure' of a group that contains instances. Do you think that would be worth carrying over to the favorites, whether or not the group itself was favorited?

Something like this (forgive the unrealistic pipeline titles):
![Screen Shot 2021-01-29 at 2 28 24 PM](https://user-images.githubusercontent.com/6372810/106318806-737d0680-623e-11eb-86b3-acd9b81788d2.png)

aoldershaw commented:
@matthewpereira makes sense - if we chose to have allow expanding instance groups in the "all pipelines" section of the sidebar, it'd probably look similar to your mockup.

What should the badge display if both the group and instances within the group are favourited? e.g. if `a-favorited-group` had two favourited instances and was favourited itself, would we display `12` as is done now, or `2` (for consistency with unfavourited groups, at least in the mockup)?

chenbh reviewed (commented):
I'm not sure I like displaying the instance vars in the card header, it kinda looks off when there's a combination of favourited instanced and non-instanced pipelines
 
![Screen Shot 2021-02-01 at 4 11 06 PM](https://user-images.githubusercontent.com/17577073/106518603-243a0e80-64a8-11eb-8c93-0f3936ff41fa.png)

Wild ideas:
1. display the pipeline name as `pipeline/var1:value1,var2:value2`. Con: this feels like it clutters up the text
1. display the pipeline name, but also have a greyed out `value1-value2` on the same line. Con: might not be immediately obvious what the extra text means
1. add the instanced group badge and show the instance vars when you hover over the name. Con: I can't tell at a glance what the pipeline actually refers to

aoldershaw commented:
@chenbh feel free to disagree, but IMO things don't look *so* wonky when there's a single extra row in the header (currently one instance var) - my screenshot in the PR description shows an example of that. What if we followed @matthewpereira's suggestion for the sidebar and displayed all instance vars underneath the group name

<img width="1181" alt="Screen Shot 2021-02-02 at 10 23 37 AM" src="https://user-images.githubusercontent.com/26583442/106621607-b857b480-6540-11eb-9966-3799ce38cce0.png">

It all fits in this cherry-picked example with short instance var names/values, but I imagine having more than ~2 instance vars will be rare. If we find that this often overflows, perhaps we can re-evaluate.

It's tricky because there's not much horizontal space to work with in the pipeline cards, but at least separating the instance vars from the name affords us some more space.

chenbh commented:
@aoldershaw I like that design, it's a great compromise between too much vertical space and too little horizontal space.

matthewpereira commented:
@aoldershaw For now, I'd recommend leaving out expand/collapse functionality, leaving out indenting, and just changing the icon to differentiate pipeline instances. My goal with including the 'group' along with pipeline instances that are favorited independently is just to simplify the labeling and make it a little easier to recognize what's going on.

We could explore nested functionality in the future, but I think right now having teams expand/collapse is more than enough complexity to work with. I'd prefer to roll out the most simple option and then see how that tests.

aoldershaw commented:
@chenbh @matthewpereira I've updated the sidebar to use the `/` icon and fixed a few edge-case bugs I noticed (header height issues with instanced pipelines w/ no instance vars and no other favorited instances, and a reordering bug when the first instance is archived). Still have yet to address https://github.com/concourse/concourse/pull/6465#issuecomment-771714672, though

aoldershaw commented:
@chenbh update - I have now addressed https://github.com/concourse/concourse/pull/6465#issuecomment-771714672, so now instance vars are displayed in the same row in the favourites section when *not* viewing an instance group:

<img width="593" alt="Screen Shot 2021-02-10 at 5 54 58 PM" src="https://user-images.githubusercontent.com/26583442/107583569-1895c880-6bc9-11eb-9acf-3bdf60eb1b0c.png">

When viewing an instance group, however, the instance vars are allowed to span multiple rows, e.g.
<img width="627" alt="Screen Shot 2021-02-10 at 5 55 55 PM" src="https://user-images.githubusercontent.com/26583442/107583654-3b27e180-6bc9-11eb-8912-17cd6b22e534.png">

The thought behind this is that instances within a group should typically all have the same set of instance vars, so there shouldn't be empty space in the header and things will look uniform, so we might as well take advantage of the extra vertical space.

Let me know what you think

chenbh reviewed (approved):
lgtm!
